### ISSUE TITLE
Web loads outdated assets after updated from the store

### ISSUE THREAD

**Issue by TheOneRing** [owncloud/core#40167]
### Steps to reproduce
1. Install owncloud 10 docker 
2. Enable web
3. Update web from the market place

### Expected behaviour
Tell us what should happen
I'd expect the new updated version to be available
### Actual behaviour
Tell us what happens instead
The old assets shipped as part of the docker are used, the web client is not updated.

@kulmann you have a better grasp of the issue, can you fill in your assumptions?

**Comment by kulmann** [comment]
Thank you for starting the ticket! 👍 

So, if an app is installed in two different versions (one delivered with the core bundle, located in `/apps`, one installed from marketplace, located in `/apps-external` or `/custom` or whatever configuration for additional apps you have) oc10 always serves the one with the higher version. In the scenario you described, oc10 serves the app that was installed later via marketplace. Debugging that shows that this also works for `web` for the app itself (controllers etc).

It seems like in oc10+apache configurations where an url rewrite is active that removes `index.php` from the url the statement above is not true for e.g. JavaScript and CSS assets. Means: an URL like `https://owncloud.test/apps/web/js/web-app-files.js` is not served via (in case of the web app) the asset serving controller of the app, but instead by Apache directly. And - of course - Apache serves it from the path. Which is **always** the app installed in `/apps/web` and not the one with the more recent app version, as Apache doesn't know anything about app versions.

**Comment by kulmann** [comment]
The same applies on a bare metal installation where the apache config rewrites the `index.php` from the URL.

**Comment by micbar** [comment]
Qualifying it as p2, the new webUI is an important feature for oc10 

@pmaier1 @tbsbdr Can you organize that this will be handled in the support flow?

**Comment by jvillafanez** [comment]
If the urls are generated in the server, they should contain the right path, this means they should point to `/custom` if the latest version of the app is there. There could be some data in cache that isn't properly cleaned up and might generate a wrong path, but otherwise the code should work.

If they're generated client-side it will be problematic. As said, for whatever reason, the server generates the url with `/apps`, `/custom` or `/apps-external` based on the location of the latest version. Changing the behavior so any request to `/apps` will use the latest app version might be complex (need to check but seems delicate at least).
In addition, there are quite a bunch of files that aren't being handled by ownCloud (https://github.com/owncloud/core/blob/817f54f53d64249166feaa0eb9231c50ca1ed7e0/lib/private/Setup.php#L500). Taking into account that it's part of the apache's configuration that we're enforcing, it seems a planned behavior.

**Comment by jvillafanez** [comment]
I think this will depend on who needs to handle the app's assets. As said, right now they're handled by apache, so it needs to know the actual FS path to get the assets, that's why we need to provide the `/custom` or `/apps-external` paths depending on the app's actual location.
If we need to handle the assets ourselves, we'd need to create a custom symfony route for that and we'd also need to manage the browser cache and maybe other things that apache is currently doing by itself. In addition, we'd need to ensure all the paths to access the assets are generated with the `/apps` prefix, not with the (possible) `/custom` prefix.
Note that if we need to handle the assets, there will be a performance drop because we have to execute code which otherwise wouldn't be needed. Not sure by how much we'll be slowed down, maybe it won't be noticeable.

As an alternative, we could provide some kind of app discovery endpoint so it's possible to know what url should be build to access to specific apps.
For example, `/settings/apps/discovery` could return something like `{"files": {"url": "/apps/files", "version": "1.1"}, "user_ldap": {"url": "/custom/user_ldap", "version": "0.16.0"}}`

**Comment by kulmann** [comment]
thanks for the insights @jvillafanez ! 🚀 

In fact, the web app already has a controller which serves static assets, see https://github.com/owncloud/web/blob/master/packages/web-integration-oc10/lib/Controller/FilesController.php#L74

As long as you don't rewrite URLs to not contain `index.php` this controller is properly reached by most `web` app related requests. And that's exactly what this issue is about. When `index.php` is not present in the URL, apache / the webserver takes over serving those files and (of course) has no knowledge about app version related routing.

One way to fix it for the web app could in fact be to announce the correct paths for the static web assets, so that the webserver can actually serve them properly. But I always liked the idea of oC 10 hiding file system internals and only exposing app related files via one single `apps` route. But solving that is out of scope for the web repo.

Btw, caching is not that crucial here since web is a single page application anyway + browser caching should still work properly.

**Comment by jvillafanez** [comment]
Proposed action plan:
1. Remove https://github.com/owncloud/core/blob/817f54f53d64249166feaa0eb9231c50ca1ed7e0/lib/private/Setup.php#L500 line so the assets go through php code
2. Include symfony routes to deliver the assets. Basic patch is below (final PR will likely be different, it will probably need a proper controller):
    ```
    diff --git a/core/routes.php b/core/routes.php
    index 03c59b72ce..1e435f1c8d 100644
    --- a/core/routes.php
    +++ b/core/routes.php
    @@ -148,6 +148,30 @@ $this->create('heartbeat', '/heartbeat')->action(function () {
            \OC::$server->getSessionCryptoWrapper()->refreshCookie(\OC::$server->getConfig(), $expire);
     });
 
    +$this->create('apps.assets', '/apps/{app}/{path}.{ext}', [], ['app' => '[^/]+', 'path' => '.+$', 'ext' => '(css|js|svg|gif|png|html|ttf|woff|ico|jpg|jpeg|json|properties)'])->action(static function ($urlParams) {
    +  $app = $urlParams['app'];
    +  $path = $urlParams['path'];
    +  $ext = $urlParams['ext'];
    +  $appPath = \OC_App::getAppPath($app);
    +  if ($appPath !== false) {
    +    $f = "$appPath/$path.$ext";
    +    $detector = \OC::$server->getMimeTypeDetector();
    +    $mime = $detector->detect($f);
    +    header("Content-Type: $mime");
    +    header('Content-Length: ' . \filesize($f));
    +    readfile($f);
    +  }
    +});
    +$this->create('core.assets', '/core/{path}.{ext}', [], ['path' => '.+$', 'ext' => '(css|js|svg|gif|png|html|ttf|woff|ico|jpg|jpeg|json|properties)'])->action(static function ($urlParams) {
    +  $path = $urlParams['path'];
    +  $ext = $urlParams['ext'];
    +  $f = \OC::$SERVERROOT . "/core/$path.$ext";
    +  $detector = \OC::$server->getMimeTypeDetector();
    +  $mime = $detector->detect($f);
    +  header("Content-Type: $mime");
    +  header('Content-Length: ' . \filesize($f));
    +  readfile($f);
    +});
     /**
      * Asserts whether rewriting private and public links to the ownCloud Web UI is enabled.
      *

    ```
3. Adjust the `array_paths` config option. The `url` part will be obsolete because every access to any app will go through the `/apps` endpoint.
4. Check that any outgoing url doesn't contain any `/custom` or `/apps-external` endpoint. They should go through `/apps`.

In theory, the apps can overwrite the route provided by core in case there is something else to do. This means that the code handling the route in the web app shouldn't need to be changed.

**Comment by jvillafanez** [comment]
Discarding https://github.com/owncloud/core/issues/40167#issuecomment-1168581907 for now...

Maybe the problem is with the manual setup. I mean, the manual steps described in https://owncloud.dev/clients/web/backend-oc10/ will use the app from the `/apps` directory (usually, I guess, assuming the app is there), but the new version of the app will end up in `/custom`. However, the setup still points to the app in the `/apps` directory.

Will making the setup fully automatic fix this issue? Assuming we can run a custom repair step within the app for the automatic setup, the automatic steps could be:
1. Generate or update a custom oauth2 client in the oauth2 app. In particular, the `<web-url>/oidc-callback.html` point should be setup properly. The `urlgenerator->getAbsoluteUrl()` function should work to get the absolute url.
2. The oauth2 endpoint generating the client (`POST /apps/oauth2/clients`) should return enough information for the next steps, in particular, the client identifier, url and authUrl.
3. Generate the config.json automatically with all the info gathered.

Some additional things to consider:
* If the oauth2 app isn't enabled, the web app won't be enabled. An error should be visible when enabling the web app if this is the case.
* The web app will need to know the location of the oauth2 app. This is because the endpoint to generate the oauth2 client might be `/custom/oauth2/clients` or `/apps/oauth2/clients` (or maybe others) depending on the location of the app. `$appManager->getAppWebPath(...)` should do the trick.
* The web app will need some kind of repair step where this auto-setup will run. This should ensure that, when the app updates, the setup is also updated with the latest info, in particular possible location changes. Note that this should be run only once per update so the admin can manually fix any error we could make.

That might be enough, assuming the web app doesn't give additional problems with loading the assets. I mean, if the app is loaded from the `/custom` directory, the assets should be loaded from there.

**Comment by jvillafanez** [comment]
There is also a problem with the config.json file, at least in my docker environment.
Part of the apache's rewrite trace:
```
[Mon Jul 04 11:18:38.972420 2022] [rewrite:trace4] [pid 1183] mod_rewrite.c(483): [client 10.0.2.27:60366] 10.0.2.27 - - [server/sid#7fd86b9f20b8][rid#7fd86d4020a0/initial] [perdir /var/www/owncloud/] RewriteCond: input='/var/www/owncloud/custom/web/config.json' pattern='!\\.(css|js|svg|gif|png|html|ttf|woff|ico|jpg|jpeg|json|properties)$' => not-matched
[Mon Jul 04 11:18:38.972422 2022] [rewrite:trace1] [pid 1183] mod_rewrite.c(483): [client 10.0.2.27:60366] 10.0.2.27 - - [server/sid#7fd86b9f20b8][rid#7fd86d4020a0/initial] [perdir /var/www/owncloud/] pass through /var/www/owncloud/custom/web/config.json
```

The problem seems to be with the .htaccess file:
```
  RewriteCond %{REQUEST_FILENAME} !\.(css|js|svg|gif|png|html|ttf|woff|ico|jpg|jpeg|json|properties)$
  RewriteCond %{REQUEST_FILENAME} !core/img/favicon.ico$
  RewriteCond %{REQUEST_FILENAME} !/robots.txt
  RewriteCond %{REQUEST_FILENAME} !/remote.php
  RewriteCond %{REQUEST_FILENAME} !/public.php
  RewriteCond %{REQUEST_FILENAME} !/cron.php
  RewriteCond %{REQUEST_FILENAME} !/core/ajax/update.php
  RewriteCond %{REQUEST_FILENAME} !/status.php
  RewriteCond %{REQUEST_FILENAME} !/ocs/v1.php
  RewriteCond %{REQUEST_FILENAME} !/ocs/v2.php
  RewriteCond %{REQUEST_FILENAME} !/updater/
  RewriteCond %{REQUEST_FILENAME} !/ocs-provider/
  RewriteCond %{REQUEST_FILENAME} !/ocm-provider/
  RewriteCond %{REQUEST_URI} !^/.well-known/(acme-challenge|pki-validation)/.*
  RewriteRule . index.php [PT,E=PATH_INFO:$1]
```
Note that .json files are expected to be delivered by apache, not ownCloud. I'm not fully sure, but it seems that apache tries to get the .json file from the `<ocdir>/custom/web/config.json` and since it doesn't exists there it delegates to ownCloud.

The additional problem is that ownCloud doesn't have such route because it expects routes starting with "/apps" not "/custom". Even if the web app is installed in "/custom", the routes.php file in the appinfo directory generates routes into "/apps/web" (in this particular case)


If the config.json file needs to be delivered by ownCloud, I'd recommend to use routes such as "/get-config" or "/config", but not something that could be confused with a static file

**Comment by IljaN** [comment]
Very old, please reopen if necessary.
